src/gui/shellwidget/shellwidget.cpp

```diff
@@ -294,7 +285,7 @@ int ShellWidget::put(const QString& text, int row, int column,
 	int cols_changed = m_contents.put(text, row, column, fg, bg, sp,
 				bold, italic, underline, undercurl);
 	if (cols_changed > 0) {
-		QRect rect = absoluteShellRect(row, column, 1, cols_changed);
+		QRect rect = absoluteShellRect(row, 0, 1, m_contents.columns());
```

Comment:
I'm unsure. On the one hand it will make single character changes significantly longer but on the other hand, we're now calling one fillRect and one drawText for what could be entire lines, so in some cases it could make performance faster. I'm not sure how to reliably test it, but am willing to if you have a suggestion.

Reply:
There are a couple of tests in shellwidgets/test. The ones starting with bench_ should run repeated benchmarks.

For example to run the bench_scroll test you can call

```
$ ctest -R bench_scroll -VV
UpdateCTestConfiguration  from :/home/raf/Code/neovim-qt/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/raf/Code/neovim-qt/build/DartConfiguration.tcl
Test project /home/raf/Code/neovim-qt/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 11
    Start 11: bench_scroll

11: Test command: /home/raf/Code/neovim-qt/build/bin/bench_scroll
11: Test timeout computed to be: 10000000
11: ********* Start testing of Test *********
11: Config: Using QtTest library 5.10.1, Qt 5.10.1 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 8.2.0)
11: PASS   : Test::initTestCase()
11: PASS   : Test::benchScrollRegion()
11: RESULT : Test::benchScrollRegion():
11:      0.42 msecs per iteration (total: 54, iterations: 128)
11: PASS   : Test::cleanupTestCase()
11: Totals: 3 passed, 0 failed, 0 skipped, 0 blacklisted, 219ms
11: ********* Finished testing of Test *********
1/1 Test #11: bench_scroll .....................   Passed    0.24 sec

The following tests passed:
	bench_scroll

100% tests passed, 0 tests failed out of 1
```

Which gives the average time for one test execution.

We could try something similar using the ShellWidget class directly. For comparison here what we would want is two tests, one with using ligatures and one without.

If you check test_shellcontents.cpp you'll see that you can load data from a file instead of doing it programaticaly.

Reply:
Something like this

```diff
diff --git a/src/gui/shellwidget/test/CMakeLists.txt b/src/gui/shellwidget/test/CMakeLists.txt
index 92ff991..7fe254b 100644
--- a/src/gui/shellwidget/test/CMakeLists.txt
+++ b/src/gui/shellwidget/test/CMakeLists.txt
@@ -15,3 +15,4 @@ add_xtest(test_shellcontents)
 add_xtest(test_shellwidget)
 add_xtest(bench_scroll)
 add_xtest(bench_cell)
+add_xtest(bench_widgetpaint)
diff --git a/src/gui/shellwidget/test/bench_widgetpaint.cpp b/src/gui/shellwidget/test/bench_widgetpaint.cpp
new file mode 100644
index 0000000..f220450
--- /dev/null
+++ b/src/gui/shellwidget/test/bench_widgetpaint.cpp
@@ -0,0 +1,23 @@
+#include <QtTest/QtTest>
+#include "shellwidget.h"
+
+#if defined(Q_OS_WIN) && defined(USE_STATIC_QT)
+#include <QtPlugin>
+Q_IMPORT_PLUGIN (QWindowsIntegrationPlugin);
+#endif
+
+class Test: public QObject
+{
+       Q_OBJECT
+private slots:
+       void benchPaint() {
+               QBENCHMARK {
+                       auto s = ShellWidget::fromFile("../test/buffer.c");
+                       s->show();
+                       s->repaint();
+               }
+       }
+};
+
+QTEST_MAIN(Test)
+#include "bench_widgetpaint.moc"
```

In this example I'm using one of the files in the repo which might not have many ligatures.

Reply:
Heads up on that example

- that path is probably wrong
- before forcing a repaint you will likely want to resize

Reply:
It actually has a decent amount of them, but I think this is more about rendering regular code anyway.
Time on master:
```
13: PASS   : Test::initTestCase()
13: PASS   : Test::benchPaint()
13: RESULT : Test::benchPaint():
13:      3.5 msecs per iteration (total: 56, iterations: 16)
13: PASS   : Test::cleanupTestCase()
13: Totals: 3 passed, 0 failed, 0 skipped, 0 blacklisted, 256ms
```

Time on ligature branch:

```
13: PASS   : Test::initTestCase()
13: PASS   : Test::benchPaint()
13: RESULT : Test::benchPaint():
13:      4.6 msecs per iteration (total: 74, iterations: 16)
13: PASS   : Test::cleanupTestCase()
13: Totals: 3 passed, 0 failed, 0 skipped, 0 blacklisted, 230ms
```

This isn't necessarily representative. I've run both a bunch of times and got higher and lower results with both. In general it would seem the new code has less consistent draw time. I'll add a few more lines to the test and try some more testing after adding the property.

Reply:
A slightly better test case. Sorry about the formatting mess. The jpg file is just so I can examine what the test ended up doing.

```c++
#include <QtTest/QtTest>
#include <QScreen>
#include "shellwidget.h"
#include "helpers.h"

#if defined(Q_OS_WIN) && defined(USE_STATIC_QT)
#include <QtPlugin>
Q_IMPORT_PLUGIN (QWindowsIntegrationPlugin);
#endif

class Test: public QObject
{
       Q_OBJECT
		ShellWidget *s;
private slots:
		void init() {
			// FIXME: path
			QFileInfo fi("../../../../../src/gui/shellwidget/test/shellcontents/shell2.txt");
			s = ShellWidget::fromFile(fi.absoluteFilePath());
			s->resizeShell(80, 120);
			s->show();
		}
		void cleanupTestCase() {
			// save a image of the widget after the last test
			s->grab().save("widgetpaint.png");
		}
       void benchPaint() {
               QBENCHMARK {
				   s->repaint();
               }
       }
};

QTEST_MAIN(Test)
#include "bench_widgetpaint.moc"
```

Reply:
Ok I've included a working benchmark with fixes for the path here https://github.com/equalsraf/neovim-qt/commit/736f57af33ca2c4d9a9ece1f0e5c1781630310db